### ISSUE TITLE
iscsid: iSCSI re-opening session attempts are causing log flooding.

### ISSUE THREAD

**Issue by shmsingh** [open-iscsi/open-iscsi#487]
When an iscsi session goes down (it can be due to ISCSI_ERR_TCP_CONN_CLOSE) iscsid attempts to reopen the session every 3 seconds. 

```
#define ISCSI_CONN_ERR_REOPEN_DELAY     3
queue_delayed_reopen(qtask, ISCSI_CONN_ERR_REOPEN_DELAY);
```

iscsid will also dump reoepn attempt logs to the console.

```
Oct 15 15:36:30 scaqap06adm08 iscsid[26768]: iscsid: Poll was woken by an alarm
Oct 15 15:36:30 scaqap06adm08 iscsid[26768]: iscsid: re-opening session 6 (reopen_cnt 7399)
Oct 15 15:36:30 scaqap06adm08 iscsid[26768]: iscsid: disconnecting conn 0x560d7e2045c0, fd 6
Oct 15 15:36:30 scaqap06adm08 iscsid[26768]: iscsid: connecting to 192.168.46.1:3260
Oct 15 15:36:30 scaqap06adm08 iscsid[26768]: iscsid: cannot make a connection to 192.168.46.1:3260 (-1,101)
Oct 15 15:36:30 scaqap06adm08 iscsid[26768]: iscsid: re-opening session 2 (reopen_cnt 7398)
Oct 15 15:36:30 scaqap06adm08 iscsid[26768]: iscsid: disconnecting conn 0x560d7e020340, fd 8
Oct 15 15:36:30 scaqap06adm08 iscsid[26768]: iscsid: connecting to 192.168.46.1:3260
Oct 15 15:36:30 scaqap06adm08 iscsid[26768]: iscsid: cannot make a connection to 192.168.46.1:3260 (-1,101) 
```

And if the session never restored or restored after very long time, these reopen logs will fill up the console log file and can potentially truncate the exisiting logs. In worst case it can truncate the important debug logs from any previously faced issue.

Is it possible to reduce the frequency of dumping these logs on the console?

**Comment by mikechristie** [comment]
There is no code/setting for this.

**Comment by gonzoleeman** [comment]
On Oct 24, 2024, at 3:12 PM, mikechristie ***@***.***> wrote:
> 
> 
> There is no code/setting for this.
> 

Does it really make sense to retry forever in this case? Perhaps this should be a (new) setting?

Settings could be: -1 (don’t retry), 0 (retry forever), INTEGER: retry that many times.

— 
Lee

**Comment by gonzoleeman** [comment]
Apologies. I didn’t mean to hijack the original issue. And without research. 
--
Lee-Man Duncan

**Comment by shmsingh** [comment]
> There is no code/setting for this.

Right. I explored the iscsid code and didn't find any. But is it possible to add one (like what @gonzoleeman  suggested). Or what all other options do we have.

**Comment by mikechristie** [comment]
There's 2 issues right?
1. Too many log messages.
2.  Retrys. In some cases it makes sense to retry forever. Some users just want it to never fail because of iscsi or the failure can last for 1 to 10+ hours so it's just easier.

For #1 we don't have anything.

For #2 is this what node.session.reopen_max is for? The iscsid.conf description mentions iscsid restarting but I think the code will limit re-logins as a result of session recovery also. @gonzoleeman I swear I think you added this code right? Maybe it just needs some tweaks.

**Comment by shmsingh** [comment]
From what I can understand reopen_max will limit the max number of retries it will peform to reopen the session. This is not what I am looking for. Problem is not with the retries but with too many logs messages that comes with it. As per your comment and what I can see in the code, currently there is no option available to limit that.

I am not talking about the existing code but referring to adding something new. My query is can we have some kind of logic in code or a control config variable to limit or reduced reopen attemp frequency of dumping these logs per connection/session based. eg. Let it keep trying attempt every 3 seconds, but for logs instead of every re-open attempt, dump on every nth attempt. Maybe we can add 2 variables, one to set/reset this feature and one to control the frerquency.
Please share your thoughts on this.

**Comment by mikechristie** [comment]
Like I wrote in the email to your work address, feel free to send a patch.

I don't think you need 2 settings. One can probably convey the needed info like an int where < 0 means off and > 0 means the freqquency. Although I don't know what you mean by reset the feature.

**Comment by biger410** [comment]
Mike, to be clear, are you suggesting a new configuration that will control retry, 0 means no retry, and non zero means how many seconds retry once?

**Comment by gonzoleeman** [comment]
On Oct 31, 2024, at 9:48 AM, Junxiao Bi ***@***.***> wrote:
> 
> ﻿
> Mike, to be clear, are you suggesting a new configuration that will control retry, 0 means no retry, and non zero means how many seconds retry once?
> 

I believe we already have an attribute that controls retries. I believe he meant a new attribute that controls warning log messages about those retries.
